zio/csupio/vectorreader.go

```diff
@@ -77,27 +88,26 @@ func (v *VectorReader) Pull(done bool) (vector.Any, error) {
 		v.close()
 		return nil, err
 	}
-	n := int(v.nextObject.Add(1) - 1)
-	if n >= len(v.objects) {
-		return nil, v.close()
-	}
-	o := v.objects[n]
-	return vcache.NewObjectFromCSUP(o).Fetch(v.zctx, v.projection)
-}
-
-func filterObjects(zctx *super.Context, pruner expr.Evaluator, objects []*csup.Object, projection field.Projection) []*csup.Object {
-	if pruner == nil {
-		return objects
+	for {
+		o, err := v.stream.next()
+		if o == nil || err != nil {
+			return nil, err
+		}
+		// XXX using the query context for the metadata filter unnecessarily
+		// pollutes the type context.  We should use the csup local context for
+		// this filtering but this will require a little compiler refactoring to be
+		// able to build runtime expressions that use different type contexts.
+		if v.metaFilters == nil || !pruneObject(v.zctx, v.metaFilters, o.Metadata()) {
+			return vcache.NewObjectFromCSUP(o).Fetch(v.zctx, v.projection)
+		}
 	}
-	return slices.DeleteFunc(objects, func(o *csup.Object) bool {
-		return pruneObject(zctx, pruner, o.Metadata(), projection)
-	})
 }
 
-func pruneObject(zctx *super.Context, pruner expr.Evaluator, m csup.Metadata, projection field.Projection) bool {
-	vals := csup.ProjectMetadata(zctx, m, projection)
+func pruneObject(zctx *super.Context, metaFilters *sync.Pool, m csup.Metadata) bool {
+	mf := metaFilters.Get().(*metafilter)
```

Comment:
Put back the metafilter once you are done?

Reply:
oopsies